### PR TITLE
change hash algorithm compatible for fetchNextcloudApp

### PR DESCRIPTION
**Pull request by onny**

### CODE REVIEW

**Comment by Scriptkiddi** [comment]
Hey isnt there a better way to calculate that hash instead of spawing seperate processes? And why is this a good idea?

**Comment by onny** [comment]
@Scriptkiddi I guess calculating hashes has changed since https://github.com/NixOS/nixpkgs/pull/193075 and nc4nix probably didn't got updated yet.
Yeah it would be cool to calculate it with Go but not sure how to do it

**Comment by Scriptkiddi** [comment]
I think we should rework this to go code and then merge it

**Comment by onny** [comment]
> I think we should rework this to go code and then merge it

Found a way to calculate the hash in Go without depending on external commands :)

**Comment by onny** [comment]
@Scriptkiddi @dasJ

File: main.go [hunk 1]
```diff
@@ -118,9 +126,30 @@ func prefetch(url string) (string, error) {
 	contents, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		log.Print("Prefetch failed reading body for: ", url, err)
+	}
+	dname, err := os.MkdirTemp("", "")
+	if err != nil {
+		log.Fatal(err)
```
Comment (leonklingele) [inline main.go:132]: Don't log the error and instead return it:

```suggestion
		return "", fmt.Errorf("failed to create temp dir: %w", err)
```

File: main.go [hunk 2]
```diff
@@ -118,9 +126,30 @@ func prefetch(url string) (string, error) {
...
 	}
-	sha256 := fmt.Sprintf("%x", sha256.Sum256(contents))
+	var entryPath string
+	for _, entry := range entries {
+		if entry.IsDir() {
+			entryPath = filepath.Join(dname, entry.Name())
+			break
+		}
+	}
+	err = nar.DumpPath(h, entryPath)
```
Comment (leonklingele) [inline main.go:147]: Combine this declaration with the next line:

```go
if err := nar.DumpPath(h, entryPath); err != nil {
```

File: main.go [hunk 3]
```diff
@@ -118,9 +126,30 @@ func prefetch(url string) (string, error) {
...
+	if err != nil {
+		panic(err)
```
Comment (leonklingele) [inline main.go:149]: Don't panic, instead return the error:

```suggestion
		return "", fmt.Errorf("failed to dump path: %w", err) 
```

File: main.go [hunk 4]
```diff
@@ -118,9 +126,30 @@ func prefetch(url string) (string, error) {
...
+	}
+	defer os.RemoveAll(dname)
+	sha256 = nixbase32.EncodeToString(h.Sum(nil))
```
Comment (leonklingele) [inline main.go:152]: Why not remove the `var sha256 string` above and make this a

```suggestion
	return nixbase32.EncodeToString(h.Sum(nil)), nil
```

**Review by leonklingele — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
Thank you @leonklingele , fixed the issues 👍

File: main.go [hunk 5]
```diff
@@ -109,6 +115,8 @@ func writeFile(t string, c AppJson) {
 }
 
 func prefetch(url string) (string, error) {
+	h := sha256.New()
+	var sha256 string
```
Comment (leonklingele) [inline main.go:119]:
```suggestion
```

**Comment by onny** [comment]
Oh missed that, thank you!

**Review by leonklingele — APPROVED** [review]

**Comment by onny** [comment]
@ajs124

**Comment by onny** [comment]
Is there anything I can test or further add please let me know :)

**Comment by leonklingele** [comment]
@onny thanks! Code-wise this lgtm, what's the intention behind this change though? Does the current version (i.e. latest `master`) yields any different (incorrect?) hashes?

**Comment by onny** [comment]
yes the hash is now different, probably because fetchnextcloudapp requires a recursive hash, not of the single zip file line before, but of the zip content recursively https://github.com/NixOS/nixpkgs/blob/c5e62ec76a0198822c1cf017f83c1d4c9d1ceb60/pkgs/servers/nextcloud/packages/default.nix#L18

**Comment by leonklingele** [comment]
To me this change looks like it's switching to the new hash _representation_ used by nix which additionally allows for different hashing algorithms (e.g. `sha1`, .. instead of `sha256`). i.e., it's not _changing_ the hash, it's merely using a _different encoding_ of it.

@Conni2461 @dasJ what do you think? nc4nix is more performant when not using this change, and instead leaving the hash-calculation to stdlib's `sha256.Sum256`.

**Comment by onny** [comment]
> To me this change looks like it's switching to the new hash _representation_ used by nix which additionally allows for different hashing algorithms (e.g. `sha1`, .. instead of `sha256`). i.e., it's not _changing_ the hash, it's merely using a _different encoding_ of it.
> 
> @Conni2461 @dasJ what do you think? nc4nix is more performant when not using this change, and instead leaving the hash-calculation to stdlib's `sha256.Sum256`.

sha256.Sum256 only hashes the zip file whereas this PR calculates the hash by unpacking the zip and calculating the hash recursively over all containing files, see:
```
$ nix-prefetch-url --unpack "https://github.com/nextcloud/bookmarks/releases/download/v13.1.3/bookmarks-13.1.3.tar.gz"
06pprhlaaqdha2nmfdcf76mhh48hdr5jlv88snxji8lpflv50wr5
$ nix-prefetch-url  "https://github.com/nextcloud/bookmarks/releases/download/v13.1.3/bookmarks-13.1.3.tar.gz"
0yl1agr62y87bil321ryrl0qwp9m0gksgwhvcqqlkxgwnjs98d7n
```

the unpacked hash is used in fetchNextcloudApp because this function also uses fetchzip with recursive hash calculation: https://github.com/NixOS/nixpkgs/blob/791355b27c5bfc58f7339a3dafc936ec50130e01/pkgs/build-support/fetchnextcloudapp/default.nix

**Comment by leonklingele** [comment]
> sha256.Sum256 only hashes the zip file whereas this PR calculates the hash by unpacking the zip and calculating the hash recursively over all containing files

I'm not 100% fluent in Nix but I can't see a difference here from a integrity-preserving point of view. Let's see what my work mates have to say. I'll get back to you tomorrow! :)

File: main.go [hunk 6]
```diff
@@ -118,11 +125,30 @@ func prefetch(url string) (string, error) {
 	contents, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		log.Print("Prefetch failed reading body for: ", url, err)
+	}
+	dname, err := os.MkdirTemp("", "")
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp dir: %w", err)
+	}
```
Comment (leonklingele) [inline main.go:132]: Just noticed that this directory should be removed after use here, and not further down below, i.e.

```suggestion
	dname, err := os.MkdirTemp("", "")
	if err != nil {
		return "", fmt.Errorf("failed to create temp dir: %w", err)
	}
	defer func() {
		if err := os.RemoveAll(dname); err != nil {
			log.Printf("failed to clean up temp dir %q: %v", dname, err)
		}
	}()
```

File: main.go [hunk 7]
```diff
@@ -118,11 +125,30 @@ func prefetch(url string) (string, error) {
...
+	for _, entry := range entries {
+		if entry.IsDir() {
+			entryPath = filepath.Join(dname, entry.Name())
+			break
+		}
+	}
+	if err := nar.DumpPath(h, entryPath); err != nil {
+		return "", fmt.Errorf("failed to dump path: %w", err)
+	}
+	defer os.RemoveAll(dname)
```
Comment (leonklingele) [inline main.go:149]:
```suggestion
```

**Review by leonklingele — CHANGES_REQUESTED** [review]
Sorry, this slipped!

**Comment by leonklingele** [comment]
My work mates are busy this week, unfortunately. We‘ll get back to you asap!

**Comment by cheriimoya** [comment]
I checked the output and somehow, there are some empty values. If I run `rg '"sha256": ""' *.json` but I just noticed, it also happens with the current implementation. Maybe those should be handled/debugged in a separate Issue.

Otherwise, the PR looks good to me and can be merged.

**Comment by onny** [comment]
thank you so much for helping and merging this :)

**Comment by leonklingele** [comment]
Hey @onny :) Just fyi: We have updated this again with https://github.com/helsinki-systems/nc4nix/commit/3c35a0efda57bb633a137463f46266ba3b04d3c0 — we now no longer extract the zip archive due to bug with some of the Nextcloud apps which yielded an incorrect hash. We however still use an SRI hash.

**Comment by dotlambda** [comment]
> We have updated this again with [3c35a0e](https://github.com/helsinki-systems/nc4nix/commit/3c35a0efda57bb633a137463f46266ba3b04d3c0) — we now no longer extract the zip archive due to bug with some of the Nextcloud apps which yielded an incorrect hash.

Does that not create problems regarding reproducability as outlined in https://github.com/NixOS/nixpkgs/pull/193075?
cc @Ma27

**Pull request merged**